### PR TITLE
Sweets/erc6551

### PR DESCRIPTION
**Pull request by SweetmanTech**
# ERC6551 - token bound accounts
- uses `_afterTokenTransfers` hook. 🪝 
- only register for mints when `from == address(0)`. ✨ 
- only register if ERC6551 registry has been set when `erc6551Registry != address(0)`. ✨ 
- includes Unit Tests. 🧪 
Screenshot 2023-07-12 at 12 18 37 PM

### Unit Tests 🧪 
- `test_Erc6551Registry` - verifies initial ERC6551Registry has been deployed. 🧪 
- `test_createAccount` - verifies `purchase` automatically creates a token bound account. 🧪 
- `test_setErc6551Registry_revert_Access_OnlyAdmin` - verifies `onlyAdmin` can call `setErc6551Registry`. 🧪 
- `test_setErc6551Implementation_revert_Access_OnlyAdmin` - verifies `onlyAdmin` can call `setErc6551Implementation`. 🧪 
- `test_createMultipleAccounts` - verifies minting `multiple` CRE8ORS registers ERC6551 TBAs for all tokens. 🧪 
- `test_sendWithTBA` - Send PFP to TBA, verify ownership, remove from TBA, verify no more ownership. 🧪 
- `test_sendWithTBA_revert_NotAuthorized` - verifies TBA can only be used for signing if the caller owns the nft. 🧪 

### Special Unit Test
- `test_transfer_only` - Transfer (not mint) does NOT call `register`. 🧪 
- compare the `gas` to see the gas savings by removing registration on transfers. ⛽ 
Screenshot 2023-07-12 at 5 23 23 PM

### ERC6551
Learn More: https://eips.ethereum.org/EIPS/eip-6551
Screenshot 2023-07-12 at 12 18 10 PM

### CODE REVIEW

File: src/interfaces/IAccountGuardian.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    function defaultImplementation() external view returns (address);
+
+    function isTrustedImplementation(
+        address implementation
+    ) external view returns (bool);
+
+    function isTrustedExecutor(
+        address implementation
+    ) external view returns (bool);
+}
```
Comment (SweetmanTech) [inline src/interfaces/IAccountGuardian.sol:21]: library copied from ERC6551 repo.

File: src/interfaces/IERC6551Account.sol [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    function token()
+        external
+        view
+        returns (uint256 chainId, address tokenContract, uint256 tokenId);
+
+    function owner() external view returns (address);
+
+    function nonce() external view returns (uint256);
+}
```
Comment (SweetmanTech) [inline src/interfaces/IERC6551Account.sol:32]: library copied from ERC6551 repo.

https://docs.tokenbound.org/contracts/registry

File: src/interfaces/IERC6551Registry.sol [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    ) external returns (address);
+
+    function account(
+        address implementation,
+        uint256 chainId,
+        address tokenContract,
+        uint256 tokenId,
+        uint256 salt
+    ) external view returns (address);
+}
```
Comment (SweetmanTech) [inline src/interfaces/IERC6551Registry.sol:30]: library copied from ERC6551 repo.

https://docs.tokenbound.org/contracts/registry

File: src/utils/AccountGuardian.sol [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    function setTrustedExecutor(
+        address executor,
+        bool trusted
+    ) external onlyOwner {
+        isTrustedExecutor[executor] = trusted;
+        emit TrustedExecutorUpdated(executor, trusted);
+    }
+}
```
Comment (SweetmanTech) [inline src/utils/AccountGuardian.sol:32]: library copied from ERC6551 repo.

https://docs.tokenbound.org/contracts/registry

File: src/utils/ERC6551AccountLib.sol [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        bytes memory footer = new bytes(0x20);
+
+        assembly {
+            // copy 0x20 bytes from beginning of footer
+            extcodecopy(address(), add(footer, 0x20), 0x2d, 0x4d)
+        }
+
+        return abi.decode(footer, (uint256));
+    }
+}
```
Comment (SweetmanTech) [inline src/utils/ERC6551AccountLib.sol:26]: library copied from ERC6551 repo.

https://docs.tokenbound.org/contracts/registry

File: src/utils/ERC6551BytecodeLib.sol [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    ) internal pure returns (bytes memory) {
+        return
+            abi.encodePacked(
+                hex"3d60ad80600a3d3981f3363d3d373d3d3d363d73",
+                implementation_,
+                hex"5af43d82803e903d91602b57fd5bf3",
+                abi.encode(salt_, chainId_, tokenContract_, tokenId_)
+            );
+    }
+}
```
Comment (SweetmanTech) [inline src/utils/ERC6551BytecodeLib.sol:20]: library copied from ERC6551 repo.

https://docs.tokenbound.org/contracts/registry

File: src/utils/ERC6551Registry.sol [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                chainId,
+                tokenContract,
+                tokenId,
+                salt
+            )
+        );
+
+        return Create2.computeAddress(bytes32(salt), bytecodeHash);
+    }
+}
```
Comment (SweetmanTech) [inline src/utils/ERC6551Registry.sol:72]: library copied from ERC6551 repo.

https://docs.tokenbound.org/contracts/registry

File: src/Cre8ors.sol [hunk 8]
```diff
@@ -418,6 +420,36 @@ contract Cre8ors is
...
+    function _afterTokenTransfers(
+        address from,
+        address,
+        uint256 startTokenId,
+        uint256 quantity
+    ) internal override {
+        if (from == address(0) && erc6551Registry != address(0)) {
+            createTokenBoundAccounts(startTokenId, quantity);
+        }
+    }
```
Comment (SweetmanTech) [inline src/Cre8ors.sol:437]: the main magic happens here to spin up an ERC6551 🪄

File: src/minter/BurnCrosschainMinter.sol [hunk 9]
```diff
@@ -61,11 +63,10 @@ contract BurnCrosschainMinter {
...
     ) external payable returns (uint256) {
-        (
-            bytes32 hashedFrom,
-            uint256 burnQuantity,
-            bytes32 encryptedPasscode
-        ) = abi.decode(data, (bytes32, uint256, bytes32));
+        (, uint256 burnQuantity, bytes32 encryptedPasscode) = abi.decode(
+            data,
+            (bytes32, uint256, bytes32)
+        );
```
Comment (SweetmanTech) [inline src/minter/BurnCrosschainMinter.sol:69]: fixing linter warnings for unused variables 🚮

File: src/utils/Cre8orsERC6551.sol [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ ██████╗██████╗ ███████╗ █████╗  ██████╗ ██████╗ ███████╗
+██╔════╝██╔══██╗██╔════╝██╔══██╗██╔═══██╗██╔══██╗██╔════╝
+██║     ██████╔╝█████╗  ╚█████╔╝██║   ██║██████╔╝███████╗
+██║     ██╔══██╗██╔══╝  ██╔══██╗██║   ██║██╔══██╗╚════██║
+╚██████╗██║  ██║███████╗╚█████╔╝╚██████╔╝██║  ██║███████║
+ ╚═════╝╚═╝  ╚═╝╚══════╝ ╚════╝  ╚═════╝ ╚═╝  ╚═╝╚══════╝                                                       
+ */
+/// @dev inspiration: https://github.com/ourzora/zora-drops-contracts
+contract Cre8orsERC6551 {
```
Comment (SweetmanTech) [inline src/utils/Cre8orsERC6551.sol:15]: add a `constructor` to set initial values of
- `erc6551Registry`
- `erc6551AccountImplementation`

File: src/interfaces/IERC6551Registry.sol (hunk 3 above)
Comment (crazzywizard) [reply src/interfaces/IERC6551Registry.sol:30]: was it not possible to install these as git submodules from https://github.com/erc6551/reference? just curious.

File: test/ERC6551.t.sol [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+            address(cre8orsNFTBase),
+            value,
+            data
+        );
+        uint256 tokenId = abi.decode(response, (uint256));
+        assertEq(cre8orsNFTBase.balanceOf(tokenBoundAccount), 1);
+
+        // use TBA to burn CRE8OR
+        data = abi.encodeWithSignature(
+            "safeTransferFrom(address,address,uint256)",
```
Comment (crazzywizard) [inline test/ERC6551.t.sol:129]: so you have to give it a function interface definition in order for the account to perform actions?

**Review by crazzywizard — APPROVED** [review]
LGTM!

File: src/interfaces/IERC6551Registry.sol (hunk 3 above)
Comment (SweetmanTech) [reply src/interfaces/IERC6551Registry.sol:30]: testing 🧪

File: test/ERC6551.t.sol [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+        vm.prank(DEFAULT_OWNER_ADDRESS);
+        cre8orsNFTBase.purchase(3);
+        assertTrue(isContract(tokenBoundAccount));
+
+        // TBA used to mint another CRE8OR
+        assertEq(cre8orsNFTBase.balanceOf(tokenBoundAccount), 0);
+        uint256 value;
+        bytes memory data = abi.encodeWithSignature("purchase(uint256)", 1);
+        vm.expectRevert(NotAuthorized.selector);
+        Account(tokenBoundAccount).executeCall(
```
[the remark this answers is not in the document]
Comment (SweetmanTech) [reply test/ERC6551.t.sol:156]: Only the token `ownerOf` or, the token itself, can call `executeCall` from the ERC6551 wallet.

Here, the `ownerOf(1)` is `DEFAULT_OWNER_ADDRESS`. On `line 147`, we `prank` `DEFAULT_OWNER_ADDRESS` for only the call to `purchase`. After that, it's a random wallet calling the following tx.

on `line 156` it's a random address calling `executeCall`. The random address does not own the `tokenBoundAccount`, so, the TX reverts, as expected.

Does that answer your question?

File: src/interfaces/IERC6551Registry.sol (hunk 3 above)
Comment (SweetmanTech) [reply src/interfaces/IERC6551Registry.sol]: updated ✅

Comment (SweetmanTech) [reply src/interfaces/IERC6551Registry.sol]: added 2 new submodules
- `lib/ERC6551` - Registry implementations. ✨ 
- `lib/tokenbound` - Account implementations. ✨ 
Screenshot 2023-07-14 at 12 07 42 PM

File: test/ERC6551.t.sol (hunk 11 above)
Comment (SweetmanTech) [reply test/ERC6551.t.sol:129]: exactly! That way, the `Account` contract can be very simple, while still having the composability with ANY contract call in the Ethereum ecosystem. 🌈

**Pull request merged**